core/src/tile/tileWorker.cpp

```diff
@@ -30,11 +33,17 @@ TileWorker::~TileWorker(){
     }
 }
 
-void TileWorker::run() {
+void disposeBuilder(std::unique_ptr<TileBuilder> _builder) {
+    if (_builder) {
+        Tangram::runOnMainLoop([b = std::shared_ptr<TileBuilder>(std::move(_builder))](){});
```

Comment:
The shared_ptr is stored in the function (to which this lamba gets converted). The function runs on the next mainloop iteration and so the last shared_ptr is destructed with it.
std::function must be copyable, that's why _builder is upgraded to a shared_ptr. 

I think a compiler is not allowed to optimize any of this away.